### PR TITLE
Add support for GL_EXT_blend_func_extended in GLES2 renderer

### PR DESCRIPTION
**Pull request by anarsoul**
GLES2 has GL_EXT_blend_func_extended extension that enables
dual-source blending, so essentially we can use fragment shader
that is very similar to GLSL3 renderer and do 1 rendering pass
instead of 3 for the text.

Tested on Pinebook with lima driver (ARM Mali 400 GPU)

### CODE REVIEW

**Comment by anarsoul** [comment]
I'm a pretty fresh rustacean, so please don't beat me too hard :) Any review is welcome and I'll be more than happy to address the comments.

**Comment by kchibisov** [comment]
>GLES2 has GL_EXT_blend_func_extended extension that enables
dual-source blending, so essentially we can use fragment shader
that is very similar to GLSL3 renderer and do 1 rendering pass
instead of 3 for the text.

gles2 doesn't have this extension, this is gles3 thing. Also the amount of sources should be the same as in glsl3 shader.

**Comment by kchibisov** [comment]
Also, if you want to make use of it you should change glsl3 renderer and its shader, no need to add new shader which is the same as glsl3 shader.

**Comment by anarsoul** [comment]
> gles2 doesn't have this extension, this is gles3 thing.

See https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_blend_func_extended.txt

```
<...>
This extension is written against the OpenGL ES 3.1 (June 4, 2014)
    specification, but can apply to earlier versions back to ES 2.0.
<...>
 GLSL version 100 language is written against The OpenGL ES Shading
    Language (May 12, 2009).
<...>
```

Moreover, I'm the guy who implemented dual-source blending in lima driver, so I can tell you form my personal experience that this extension can be used with GLES2. Mali4x0 is GLES2 class GPU and it supports this extension, however I must admit that lima is the only driver for GLES2-class GPUs I know that exposes this extension.

>  Also the amount of sources should be the same as in glsl3 shader.

Fragment shader is very similar (and actually it's rewritten GLSL3 shader), but ESSL 1.0 and GLSL 3.3 are different enough, so I put it in a separate file. Note that ESSL 1.0 doesn't allow user-defined shader outputs, so EXT_blend_func_extended adds `gl_SecondaryFragColor_EXT`

**Comment by anarsoul** [comment]
> Also, if you want to make use of it you should change glsl3 renderer and its shader, no need to add new shader which is the same as glsl3 shader.

Mali4x0 isn't capable of GL 3.3, but it exposes EXT_blend_func_extended with lima driver. Basically adding support for this extension to glsl3 defeats the purpose for me.

**Comment by kchibisov** [comment]
>Fragment shader is very similar (and actually it's rewritten GLSL3 shader), but ESSL 1.0 and GLSL 3.3 are different enough, so I put it in a separate file. Note that ESSL 1.0 doesn't allow user-defined shader outputs, so EXT_blend_func_extended adds gl_SecondaryFragColor_EXT

You can take a look at rect shader to get inspiration how to tweak glsl3 fragment shader.

Also, do you have blend_func_extended, but not instance rendering, so you can't use glsl3 renderer?

>Mali4x0 isn't capable of GL 3.3, but it exposes EXT_blend_func_extended with lima driver. Basically adding support for this extension to glsl3 defeats the purpose for me.

What extensions lima driver has?

**Comment by anarsoul** [comment]
> Also, do you have blend_func_extended, but not instance rendering, so you can't use glsl3 renderer?

That's correct

>  What extensions lima driver has?

```
GL_VERSION: OpenGL ES 2.0 Mesa 22.2.0-devel (git-53fe6f1084)
GL_RENDERER: Mali400
GL_EXTENSIONS:
    GL_EXT_blend_minmax, GL_EXT_multi_draw_arrays,
    GL_EXT_texture_format_BGRA8888, GL_OES_compressed_ETC1_RGB8_texture,
    GL_OES_depth24, GL_OES_element_index_uint, GL_OES_fbo_render_mipmap,
    GL_OES_mapbuffer, GL_OES_rgb8_rgba8, GL_OES_standard_derivatives,
    GL_OES_stencil8, GL_OES_texture_3D, GL_OES_texture_half_float,
    GL_OES_texture_half_float_linear, GL_OES_texture_npot,
    GL_OES_vertex_half_float, GL_OES_EGL_image, GL_OES_depth_texture,
    GL_OES_packed_depth_stencil, GL_OES_get_program_binary,
    GL_APPLE_texture_max_level, GL_EXT_discard_framebuffer,
    GL_EXT_read_format_bgra, GL_NV_pack_subimage, GL_EXT_frag_depth,
    GL_NV_fbo_color_attachments, GL_OES_EGL_image_external, GL_OES_EGL_sync,
    GL_OES_vertex_array_object, GL_ANGLE_pack_reverse_row_order,
    GL_EXT_texture_rg, GL_EXT_unpack_subimage, GL_NV_draw_buffers,
    GL_NV_read_buffer, GL_NV_read_depth, GL_NV_read_depth_stencil,
    GL_NV_read_stencil, GL_EXT_draw_buffers, GL_EXT_map_buffer_range,
    GL_KHR_debug, GL_KHR_texture_compression_astc_ldr,
    GL_NV_pixel_buffer_object, GL_OES_required_internalformat,
    GL_OES_surfaceless_context, GL_EXT_separate_shader_objects,
    GL_EXT_compressed_ETC1_RGB8_sub_texture, GL_EXT_draw_elements_base_vertex,
    GL_EXT_texture_border_clamp, GL_KHR_context_flush_control,
    GL_OES_draw_elements_base_vertex, GL_OES_texture_border_clamp,
    GL_EXT_blend_func_extended, GL_KHR_no_error,
    GL_KHR_texture_compression_astc_sliced_3d, GL_EXT_clip_control,
    GL_KHR_parallel_shader_compile, GL_MESA_bgra
```

**Comment by anarsoul** [comment]
Here's complete glxinfo output:

[glxinfo.txt](https://github.com/alacritty/alacritty/files/8747887/glxinfo.txt)

**Comment by kchibisov** [comment]
@anarsoul that makes sense, but I'd suggest to refactor glsl3 shader to be gles2 compatible the way it's done in rect renderer and its shader.

If you have troubles doing so, let me know I can try when I have time.

**Comment by anarsoul** [comment]
> I'd suggest to refactor glsl3 shader to be gles2 compatible the way it's done in rect renderer and its shader.

Sure, I can do that.

**Comment by anarsoul** [comment]
CI fails due to clippy error in alacritty/src/scheduler.rs which this PR doesn't touch. Should I fix it in this PR or I can just ignore it since apparently it's broken the same way in master?

**Comment by kchibisov** [comment]
@anarsoul would you mind if I clean up this patch? So it'll be a bit faster, given that changes would be stylistic. And the current code looks like C.

Also, when the patch was added to lima to support blend_func_extended? When I wrote this patch there wasn't support for that on pinephone, and I think it's using the same stuff as pinebook.

**Comment by anarsoul** [comment]
> would you mind if I clean up this patch? So it'll be a bit faster, given that changes would be stylistic.

Sure, go for it

> And the current code looks like C.

Yeah, it's not hard to guess why :)

> Also, when the patch was added to lima to support blend_func_extended? When I wrote this patch there wasn't support for that on pinephone, and I think it's using the same stuff as pinebook.

It's been quite a while, see https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/13873

The change should be present in mesa-22.0

**Comment by kchibisov** [comment]
@anarsoul I've refactored code a bit, would you mind verifying that it still works for you? It does work on my particular machine when forcing gles2 renderer, but testing on a pure gles2 system would be nice.

**Comment by anarsoul** [comment]
It works fine on my laptop with Intel graphics if I force GLES2 renderer, same on Pinebook with lima driver.

**Comment by anarsoul** [comment]
@kchibisov is there anything else blocking this PR from merging?

**Comment by kchibisov** [comment]
>@kchibisov is there anything else blocking this PR from merging?

It's waiting on review from @chrisduerr as you can see.

File: alacritty/res/glsl3/text.v.glsl
Comment (chrisduerr) [inline alacritty/res/glsl3/text.v.glsl:51]: Not sure what you mean by this, are bitwise operations not supported on gles2 with extension?

File: alacritty/src/renderer/mod.rs
Comment (chrisduerr) [inline alacritty/src/renderer/mod.rs:256]: This looks **very** unsafe. Are you just using `transmute` to extend the lifetime here? Because you should **definitely** not do that.

File: alacritty/src/renderer/mod.rs
Comment (chrisduerr) [inline alacritty/src/renderer/mod.rs:224]: A `Box<Iterator>` as storage on a struct seems somewhat unconventional. At that point why not just use a `Vec`?

File: alacritty/src/renderer/text/gles2.rs
Comment (chrisduerr) [inline alacritty/src/renderer/text/gles2.rs:464]:
```suggestion
    /// similar to GLES3 the renderer.
```

File: alacritty/src/renderer/text/gles2.rs
Comment (chrisduerr) [inline alacritty/src/renderer/text/gles2.rs:477]: Should use imports and fully qualified paths. `super::` just gets confusing.

**Review by chrisduerr — CHANGES_REQUESTED** [review]
Tested on my PinePhone and it seems to work without any issues.

File: alacritty/src/renderer/text/mod.rs
Comment (kchibisov) [reply alacritty/src/renderer/text/mod.rs:28]: To simplify shaders a bit, since we subtract due to lack of bitwise operations on gles2.

File: alacritty/res/glsl3/text.v.glsl
Comment (kchibisov) [reply alacritty/res/glsl3/text.v.glsl:28]: I thought that it's a bit clearer, since it's not a bit flag, but rather a packed sum. So 3 and 2 will mean that they have wide char.

File: alacritty/res/glsl3/text.v.glsl
Comment (kchibisov) [reply alacritty/res/glsl3/text.v.glsl:77]: I guess.

File: alacritty/res/glsl3/text.v.glsl
Comment (kchibisov) [reply alacritty/res/glsl3/text.v.glsl:47]: Ah, yeah. I refactored it a bit weirdly, since it wasn't a multiplier before.

File: alacritty/res/glsl3/text.v.glsl
Comment (kchibisov) [reply alacritty/res/glsl3/text.v.glsl:51]: >bitwise operations not supported on gles2 with extension

gles2 shaders don't have bitwise operations at all. It doesn't matter if you have extension or not.

File: alacritty/src/renderer/mod.rs
Comment (kchibisov) [reply alacritty/src/renderer/mod.rs:256]: This is safe, since the pointer from C library is pointing to static string by the opengl specification. So the reference will leave to the end of the program.

I'm just using what specification is tell me, if it wasn't safe I wouldn't extended it in the first place?

What would you suggest? To copy everything?

File: alacritty/src/renderer/mod.rs
Comment (kchibisov) [reply alacritty/src/renderer/mod.rs:224]: Because we don't collect? It's another iterator that handled 2 sources of extensions. it's purely for efficiency reasons, since we have 2 iterators and they both read static strings from some static storage, so having iterator like this won't allocate anything for elements used in the iterator.

By using `Vec` we'll have to collect and store everything, and there won't be a use for iterator at all. Though, we can likely refactor to some lazy initialized `HashSet<CStr>`.

File: alacritty/src/renderer/mod.rs
Comment (kchibisov) [reply alacritty/src/renderer/mod.rs:256]: Though, I think I don't really need transmute. I'll try without it, but the thing is that all the references are static here anyway.

File: alacritty/res/glsl3/text.v.glsl
Comment (chrisduerr) [reply alacritty/res/glsl3/text.v.glsl:28]: I'm mostly confused by the `HAS_` prefix tbh. I'd expect boolean prefix to either be `IS_` or nothing at all.

File: alacritty/src/renderer/mod.rs
Comment (chrisduerr) [reply alacritty/src/renderer/mod.rs]: > So the reference will leave to the end of the program.

Is that true for the Rust reference too though? Generally speaking, it seems like a **really** bad idea to extend an existing lifetime with `transmute`. Because even if the compiler doesn't use lifetime information for optimizations, at the very least it would indicate that the original lifetime is incorrect.

Comment (kchibisov) [reply alacritty/src/renderer/mod.rs]: Well there's no transmute anymore as you can see. The code is marked as `safe`...

File: alacritty/res/glsl3/text.v.glsl
Comment (kchibisov) [reply alacritty/res/glsl3/text.v.glsl:28]: That's intentional since it's not about identity, but about having, since it's a flag and not a property. So it could have `WIDE_CHAR` and `COLOR`, and it is wide char when the flag is higher than 2. For color it's colored when it's `1` or `3`, so you can't really check for `IS`, it's more about `has`.

Comment (chrisduerr) [reply alacritty/res/glsl3/text.v.glsl:28]: `has_widechar` would be more of a function name in that scenario though. For bitflags you'd name the flag itself `WIDE_CHAR` and then check if `flags.contains(WIDE_CHAR)`.

File: alacritty/res/glsl3/text.v.glsl
Comment (chrisduerr) [reply alacritty/res/glsl3/text.v.glsl:52]: Could you state this explicitly rather than just mentioning "other renderers"? It's not exactly clear what that means here.

File: alacritty/src/renderer/text/mod.rs
Comment (chrisduerr) [inline alacritty/src/renderer/text/mod.rs:32]:
```suggestion
/// Rendering passes, for both GLES2 and GLSL3 renderer.
```

File: alacritty/src/renderer/text/glsl3.rs
Comment (chrisduerr) [reply alacritty/src/renderer/text/glsl3.rs:23]: This has not been resolved.

File: alacritty/src/renderer/mod.rs
Comment (chrisduerr) [inline alacritty/src/renderer/mod.rs:257]: I think I'd prefer an `else` over an early return here.

File: alacritty/src/renderer/mod.rs
Comment (chrisduerr) [inline alacritty/src/renderer/mod.rs:232]: Global static state isn't super common in Rust code usually, but honestly for this kinda stuff I think it makes things pretty convenient. Especially because the underlying OpenGL structures are all static anyway.

File: alacritty/src/renderer/mod.rs
Comment (chrisduerr) [inline alacritty/src/renderer/mod.rs:260]: Unfortunate we have to use unsafe here, but I don't think there's any way around this? There might be some crates that manage to hide that unsafe away, but with STD I don't recall a way. Did you check?

File: alacritty/src/renderer/text/gles2.rs
Comment (chrisduerr) [inline alacritty/src/renderer/text/gles2.rs:461]: I don't think it's gramatically incorrect, but "there're" is a bit awkward. So I'd suggest changing it:

```suggestion
    /// For dual source blending, there are 2 passes; one for background, another for text,
```

File: alacritty/src/renderer/text/gles2.rs
Comment (chrisduerr) [inline alacritty/src/renderer/text/gles2.rs:409]: Wasn't changed in this PR, but it's kinda strange that we include "text rendering" twice then leave it out for the third. Might as well make it consistent so people don't think the third pass is for something else.

**Review by chrisduerr — CHANGES_REQUESTED** [review]

File: alacritty/src/renderer/mod.rs
Comment (kchibisov) [reply alacritty/src/renderer/mod.rs:260]: Yeah, the safe way is to put into Mutex.

Comment (kchibisov) [reply alacritty/src/renderer/mod.rs:260]: I'd go with `OnceCell`, since it's a part of our Cargo.lock anyway and will go into rust stdlib eventually.

File: alacritty/src/renderer/mod.rs
Comment (kchibisov) [reply alacritty/src/renderer/mod.rs:232]: Yeah, but we want a singleton here, since we expose global static immutable object.

File: alacritty/src/renderer/mod.rs
Comment (chrisduerr) [reply alacritty/src/renderer/mod.rs]: That's what I was thinking of when talking about crates, thanks!

File: alacritty/src/renderer/mod.rs
Comment (chrisduerr) [inline alacritty/src/renderer/mod.rs:235]: Moving this entire closure into a separate function would probably clean things up a lot visually.

**Review by chrisduerr — APPROVED** [review]
LGTM

File: alacritty/src/renderer/mod.rs
Comment (chrisduerr) [reply alacritty/src/renderer/mod.rs]: Decided to just do this myself real quick.

**Review by kchibisov — APPROVED** [review]
Looks good. I've also tested on macOS to ensure we're not breaking it.

**Pull request merged**
